File: javascript和css学习/quick_sort.js
//选择数组中间的值作为基准
function partition(a,first,last){
    var p = Math.floor((first+last)/2),
        left = first,
        right = last,
        temp;
    while(left <= right){
        while(left <= right && a[left] < a[p]){
            left++;
        }
        while(left <= right && a[right] > a[p]){
            right--;
        }
        if(left > right){
            break;
        }
        temp = a[left];
        a[left] = a[right];
        a[right] = temp;
        left++;
        right--;
    }
    return left;
}

function quickSortHelp(a,first,last){
    if(first >= last){
        return;
    }
    var p = partition(a,first,last);
    quickSortHelp(a,first,p-1);
    quickSortHelp(a,p,last);
}

function quickSort(a){
    quickSortHelp(a,0,a.length-1);    
}

File: quick_sort.py
def q_sort(a):
    def q_sort_helper(a,first,last):
        if first<last:
            def partition(a,first,last):
                done = 0
                left = first+1
                right = last
                while not done:
                    while left<=right and a[left] <= a[first]:
                        left += 1
                    while right>=left and a[right] > a[first]:
                        right -= 1
                    if left > right:
                        done = 1
                    else:
                        a[left],a[right] = a[right],a[left]
                        left += 1
                        right -= 1
                a[first],a[right] = a[right],a[first]
                return right
            q = partition(a,first,last)
            q_sort_helper(a,first,q-1)
            q_sort_helper(a,q+1,last)
    q_sort_helper(a,0,len(a)-1)